### ISSUE TITLE
Wedges, initializations, and sinks

### ISSUE THREAD

**Issue by falseywinchnet** [karpathy/nanochat#334]
I won't open any pull requests. 
I did the work to validate this, try it or leave it.
My attention, naively, descends on classes of problems where nanochat's cannot ever converge.

Here is my contribution: 

```py
class WedgeTransform(nn.Module):
    """Global Wedge (Symplectic Twist)."""
    def __init__(self, dim, heads):
        super().__init__()
        self.n_head = heads
        self.head_dim = dim // heads
        self.A = nn.Parameter(torch.zeros(heads, self.head_dim, self.head_dim))

    def forward(self, x):
        # x: (B, H, T, D)
        S = self.A - self.A.transpose(-1, -2) 
        flow = torch.matmul(x, S)
        return x + flow

#in attention init:
         self.wedge = WedgeTransform(d_model, n_head)
         nn.init.orthogonal_(self.W_Q.weight) #initialize Q with orthogonal.
        nn.init.normal_(self.W_K.weight, mean=0.0, std=0.02) # Initialize K nearly silent.
        nn.init.xavier_normal_(self.W_V.weight) #initalize V as normally.


#in attention forward:
        q = self.rope(self.wedge(q))
        k = self.rope(self.wedge(k))
```

This will boost your model massively. 

Lastly, the subject of attention sinks. there is a lot of discussion on this.
We found per-head scalar sinks work really well, but they are not in fused pytorch attention yet.
If you want to try our attention for yourself:

```py
import math
import copy
from dataclasses import dataclass
from typing import Optional, Tuple, List

import torch
import torch.nn as nn
import torch.nn.functional as F

class RoPE(nn.Module):
    """Rotary Positional Embeddings."""
    def __init__(self, dim, max_len=4096):
        super().__init__()
        self.dim = dim
        inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
        t = torch.arange(max_len).float()
        freqs = torch.einsum('i,j->ij', t, inv_freq)
        self.register_buffer('cos', freqs.cos())
        self.register_buffer('sin', freqs.sin())

    def forward(self, x):
        # x: (B, H, T, D)
        seq_len = x.shape[2]
        cos = self.cos[:seq_len, :].unsqueeze(0).unsqueeze(0)
        sin = self.sin[:seq_len, :].unsqueeze(0).unsqueeze(0)
        x1 = x[..., 0::2]
        x2 = x[..., 1::2]
        return torch.cat((-x2 * sin + x1 * cos, x1 * sin + x2 * cos), dim=-1)

class WedgeTransform(nn.Module):
    """Global Wedge (Symplectic Twist)."""
    def __init__(self, dim, heads):
        super().__init__()
        self.n_head = heads
        self.head_dim = dim // heads
        self.A = nn.Parameter(torch.zeros(heads, self.head_dim, self.head_dim))

    def forward(self, x):
        # x: (B, H, T, D)
        S = self.A - self.A.transpose(-1, -2) 
        flow = torch.matmul(x, S)
        return x + flow

class ConvexSoftmax(nn.Module):
    """Convex LSE (Float32 Precision)."""
    def forward(self, scores):
        m, _ = scores.max(dim=-1, keepdim=True)
        y = scores - m
        ex = y.exp()
        lse = m + ex.sum(dim=-1, keepdim=True).log()
        return torch.exp(scores - lse)


class Attention(nn.Module):
    def __init__(self, d_model, n_head, use_sink=True):
        super().__init__()
        self.d_model = d_model
        self.n_head = n_head
        self.head_dim = d_model // n_head
        self.use_sink = use_sink
        self.scale = self.head_dim ** -0.5

        # Projections
        self.W_Q = nn.Linear(d_model, d_model, bias=False)
        self.W_K = nn.Linear(d_model, d_model, bias=False)
        self.W_V = nn.Linear(d_model, d_model, bias=False)
        self.W_O = nn.Linear(d_model, d_model, bias=False)
        
        # Canon Initialization
        nn.init.orthogonal_(self.W_Q.weight)
        nn.init.normal_(self.W_K.weight, mean=0.0, std=0.02) # Silent K
        nn.init.xavier_normal_(self.W_V.weight)
        
        # Geometry/Stability components
        self.rope = RoPE(self.head_dim)
        self.wedge = WedgeTransform(d_model, n_head)
        self.softmax = ConvexSoftmax()
        
        if use_sink:
            self.sink_scalar = nn.Parameter(torch.zeros(1, n_head, 1, 1))
            self.v_null = nn.Parameter(torch.zeros(1, n_head, 1, self.head_dim))

    def forward(self, x):
        B, T, C = x.shape
        H, Dh = self.n_head, self.head_dim

        # 1. Projections
        q_raw = self.W_Q(x).view(B, T, H, Dh).transpose(1, 2) # Context Candidates
        k_raw = self.W_K(x).view(B, T, H, Dh).transpose(1, 2) # The Pulse
        v_val = self.W_V(x).view(B, T, H, Dh).transpose(1, 2) # The Content (V)

        q_geo = self.rope(self.wedge(q_raw))
        k_geo = self.rope(self.wedge(k_raw))

        Attn = (q_geo @ k_geo.transpose(-2, -1)) * self.scale
    
        mask = torch.triu(torch.ones(T, T, device=x.device), diagonal=1).bool()
        Attn.masked_fill_(mask, float('-inf'))
    
        null_scores = self.sink_scalar.expand(B, H, T, 1)
        Attn_full = torch.cat([Attn, null_scores], dim=-1)
        probs_full = self.softmax(Attn_full)
    
        out = probs_full[..., :T] @ v_val + probs_full[..., T:] * self.v_null
        out = out.transpose(1, 2).contiguous().view(B, T, C)
        return self.W_O(out)

```

**Comment by falseywinchnet** [comment]
https://colab.research.google.com/drive/1k9wVZYmYAMmB_FKvqmAeeKrnYWCd1uw4?usp=sharing 
a demo if anyone wants to play with our design

note we do not and we do not recommend norming k. this is destructive to its associative memory capacity.

**Comment by falseywinchnet** [comment]
@svlandeg @karpathy @THE ENTIRE FUCKING UNIVERSE
I am currently re-running my benchmarks
it turns out a lot of the data i based my assessment on
is faulty
i ran all my assessments on cpu and never used **model.to**
UNFORTUNANTLY
this means that the outputs are spurious and the model behavior on MPS is BAD
because for whatever internal reason it re-initializes on top of itself and has some non-linear distortions
this is a deeply disturbing situation to be in.

**Comment by falseywinchnet** [comment]
note I do use model.to in for example tiny LLM type scenarios but not toytransformer on simple ablation tasks.
so my ablation testing was yielding all kinds of flawed estimations

**Comment by falseywinchnet** [comment]
i am the biggest idiot that has ever been and i should never be allowed to contribute to anything.
```py
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import matplotlib.pyplot as plt
import itertools
from dataclasses import dataclass

# --- 1. Core Utilities ---

class RoPE(nn.Module):
    def __init__(self, dim, max_len=4096):
        super().__init__()
        self.dim = dim
        inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
        t = torch.arange(max_len).float()
        freqs = torch.einsum('i,j->ij', t, inv_freq)
        self.register_buffer('cos', freqs.cos())
        self.register_buffer('sin', freqs.sin())

    def forward(self, x):
        seq_len = x.shape[2]
        cos = self.cos[:seq_len, :].unsqueeze(0).unsqueeze(0)
        sin = self.sin[:seq_len, :].unsqueeze(0).unsqueeze(0)
        x1 = x[..., 0::2]
        x2 = x[..., 1::2]
        return torch.cat((-x2 * sin + x1 * cos, x1 * sin + x2 * cos), dim=-1)

class ConvexSoftmax(nn.Module):
    """Convex LSE (Float32 Precision)."""
    def forward(self, scores):
        m, _ = scores.max(dim=-1, keepdim=True)
        y = scores - m
        ex = y.exp()
        lse = m + ex.sum(dim=-1, keepdim=True).log()
        return torch.exp(scores - lse)

# --- 2. The Ablation Attention Block ---

class AblationAttention(nn.Module):
    def __init__(self, config, use_sink=False, use_convex=False, use_canon=False):
        super().__init__()
        self.d_model = config.n_embd
        self.n_head = config.n_head
        self.head_dim = self.d_model // self.n_head
        self.scale = self.head_dim ** -0.5
        
        self.use_sink = use_sink
        self.use_convex = use_convex
        self.use_canon = use_canon

        # Projections
        self.W_Q = nn.Linear(self.d_model, self.d_model, bias=False)
        self.W_K = nn.Linear(self.d_model, self.d_model, bias=False)
        self.W_V = nn.Linear(self.d_model, self.d_model, bias=False)
        self.W_O = nn.Linear(self.d_model, self.d_model, bias=False)

        # --- The Canon Initialization ---
        if self.use_canon:
            # "The Silent K" approach
            nn.init.orthogonal_(self.W_Q.weight)
            nn.init.normal_(self.W_K.weight, mean=0.0, std=0.02) 
            nn.init.xavier_normal_(self.W_V.weight)
        
        # Components
        self.rope = RoPE(self.head_dim)
        
        if self.use_convex:
            self.softmax = ConvexSoftmax()
            
        if self.use_sink:
            self.sink_scalar = nn.Parameter(torch.zeros(1, self.n_head, 1, 1))
            self.v_null = nn.Parameter(torch.zeros(1, self.n_head, 1, self.head_dim))

    def forward(self, x):
        B, T, C = x.shape
        H, Dh = self.n_head, self.head_dim

        # Projections
        q = self.W_Q(x).view(B, T, H, Dh).transpose(1, 2)
        k = self.W_K(x).view(B, T, H, Dh).transpose(1, 2)
        v = self.W_V(x).view(B, T, H, Dh).transpose(1, 2)

        # 1. RoPE (The Immutable Base)
        q = self.rope(q)
        k = self.rope(k)

        # 2. Attention Scores
        attn = (q @ k.transpose(-2, -1)) * self.scale
        mask = torch.triu(torch.ones(T, T, device=x.device), diagonal=1).bool()
        attn.masked_fill_(mask, float('-inf'))

        # 3. Sink Handling
        if self.use_sink:
            # Expand scalar to [B, H, T, 1]
            null_scores = self.sink_scalar.expand(B, H, T, 1)
            # Concat along the key dimension (last dim)
            attn_full = torch.cat([attn, null_scores], dim=-1)
            
            # 4. Softmax Choice
            if self.use_convex:
                probs_full = self.softmax(attn_full)
            else:
                probs_full = F.softmax(attn_full, dim=-1)
            
            # Split probs back into token-attention and sink-attention
            # probs: [B, H, T, T+1]
            probs_tokens = probs_full[..., :T]
            probs_sink = probs_full[..., T:] # [B, H, T, 1]
            
            # Weighted Sum: Tokens + Sink
            out = probs_tokens @ v + probs_sink * self.v_null
            
        else:
            # Standard Attention path
            if self.use_convex:
                probs = self.softmax(attn)
            else:
                probs = F.softmax(attn, dim=-1)
                
            out = probs @ v

        out = out.transpose(1, 2).contiguous().view(B, T, C)
        return self.W_O(out)

# --- 3. Experiment Harness ---

@dataclass
class Config:
    vocab_size: int = 64
    n_embd: int = 32
    n_head: int = 1
    n_layer: int = 1

def get_batch(batch_size=64, seq_len=48, vocab_size=64, device='cpu'):
    data = torch.randint(0, vocab_size, (batch_size, seq_len), device=device)
    half = seq_len // 2
    data[:, half:] = data[:, :half] 
    inputs = data[:, :-1]
    targets = data[:, 1:]
    return inputs, targets

def run_ablation():
    config = Config()
    VOCAB = 64
    DIM = 32
    STEPS = 800
    LR = 3e-3
    MASTER_SEED = 42

    # Define toggles
    toggles = ['Sink', 'Convex', 'Canon']
    # Generate all 8 permutations (True/False)
    # 0 = False, 1 = True
    combinations = list(itertools.product([False, True], repeat=3))
    
    results = {}
    
    print(f"--- Ablation: Sink vs Convex vs Canon ---")
    
    for combo in combinations:
        use_sink, use_convex, use_canon = combo
        
        # Generate Name
        parts = []
        if use_sink: parts.append("Sink")
        if use_convex: parts.append("Convex")
        if use_canon: parts.append("Canon")
        name = " + ".join(parts) if parts else "Baseline (RoPE Only)"
        
        print(f"Testing: {name}")

        # --- SANITIZATION ---
        torch.manual_seed(MASTER_SEED)
        
        attn_block = AblationAttention(config, use_sink, use_convex, use_canon)
        model = nn.Sequential(
            nn.Embedding(VOCAB, DIM),
            attn_block,
            nn.Linear(DIM, VOCAB)
        )
        
        # CRITICAL FIX
        model = model.to("cpu")
        
        opt = torch.optim.AdamW(model.parameters(), lr=LR)
        losses = []
        
        for step in range(STEPS):
            x, y = get_batch()
            opt.zero_grad()
            logits = model(x)
            loss = F.cross_entropy(logits.reshape(-1, VOCAB), y.reshape(-1))
            loss.backward()
            opt.step()
            losses.append(loss.item())
            
        results[name] = losses

    # Plotting
    plt.figure(figsize=(12, 8))
    
    def smooth(data):
        return np.convolve(data, np.ones(40)/40, mode='valid')

    # Sort legend by final loss
    sorted_results = sorted(results.items(), key=lambda x: np.mean(x[1][-50:]))

    for name, loss in sorted_results:
        # Highlight Baseline and Full Stack
        linewidth = 1
        alpha = 0.7
        if "Baseline" in name: 
            linewidth = 3
            color = 'black'
            linestyle = '--'
        elif "Sink + Convex + Canon" in name:
            linewidth = 3
            color = 'red'
            linestyle = '-'
        else:
            linestyle = '-'
            color = None # Auto
            
        plt.plot(smooth(loss), label=f"{name} ({np.mean(loss[-20:]):.3f})", 
                 linewidth=linewidth, linestyle=linestyle, alpha=alpha, color=color)
        
    plt.title("Attention Mechanics Ablation (Sorted by Performance)")
    plt.xlabel("Steps")
    plt.ylabel("Loss")
    plt.legend()
    plt.grid(True, alpha=0.3)
    plt.show()

if __name__ == "__main__":
    run_ablation()

```

Without any question this tells us that :
my LSE softmax doesnt hurt, at least
sinks help, like, a lot.
